timers/timers.go

```diff
@@ -0,0 +1,164 @@
+package timers
+
+import (
+	"runtime"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/uber-go/atomic"
+)
+
+const (
+	cacheline = 64
+
+	// State machine for timers.
+	stateDefault = iota
+	stateExpired
+	stateCanceled
+)
+
+type mutex struct {
+	sync.Mutex
+	// Avoid false sharing:
+	// http://mechanical-sympathy.blogspot.com/2011/07/false-sharing.html
+	pad [cacheline - unsafe.Sizeof(sync.Mutex{})]byte
+}
+
+// Timer FIXME
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	prev     *Timer
+}
+
+// Stop FIXME
+func (t *Timer) Stop() {
+	t.state.Store(stateCanceled)
```

Comment:
If you use `CAS(stateDefault, stateCanceled)`, you can return whether the timer was cancelled or not.

That will let us have correct exactly-once `call.End()` semantics